### DISCUSSION TITLE
Plugins

### DISCUSSION THREAD

**Discussion by SuddenDev** [payloadcms/payload#107] [Feature Requests & Ideas]
Hey there,
I'm pretty sure somethig like this is already on the roadmap, but do you guys plan something like plugins? I know they're a pain in the a** in Wordpress sometimes, but the fact that you can build your own one so easily is actually really handy. That's also one of the biggest drawbacks of Strapi currently imo; that their plugin and customization system is way too complicated.

What I do really like about payload is how easy to customize it is already. I think adding the possibilty to add something like community packages would also be awesome.
Let me know what you think

Cheers!

**Comment by jmikrut** [comment]
Hey @SuddenDev — 

You got it. This is definitely on the roadmap. Because of Payload's config-based functionality, many things you'd want to accomplish with Plugins can already be accomplished via reusable packages that you can plug into your configs themselves. Think.. reusable hooks, maybe custom off-the-shelf collections for common functionalities, or functions that wrap your config like how NextJS does.

For example, NextJS configs use a nice pattern that we can mimic now with Payload, like this:

```js
const config = withPlugin({
  serverURL: 'http://localhost:3000',
  collections: [
  ],
  // the rest of the config
})
```

That `withPlugin` function could take in your config, modify it however it needs, and return it with its added functionality. For example, it could:

- auto-format all of your Upload-enabled collections to add automatic S3 integration
- extend your GraphQL implementation with custom queries / mutations
- add a full set of ecommerce-enabling collections that integrate with Stripe, etc
- swap out built-in components automatically for custom components
- etc.

Super powerful pattern. "Plugins" using this approach are totally doable right now.

We plan to write up some examples like this and publish them in our docs to show how it might work - but the beautiful part here—and one that is central to Payload—is... that _**it's just JavaScript**_. No magic, no "do plugins the Payload way"...  So you definitely would never have to fight a complex plugin infrastructure. 

One personal goal of mine is to provide a config-based way for everything that a plugin might want to do to Payload, right in the config, and over time end up with a completely customizable base set of functionality that is 100% customizable.

That said, there will likely be areas of Payload that do require a deliberate "plugin" implementation. We are fleshing that out internally and it is definitely on the roadmap. More on that soon.

In the meantime, this Discussion would make for a great spot to centralize plugin ideas, as the community has them, so we can evaluate what's necessary to make them happen. If you have any ideas, send them our way and let's make it happen!

**Reply by SuddenDev** [reply]
Hey @jmikrut —

thanks for your reply!

>We plan to write up some examples like this and publish them in our docs to show how it might work - but the beautiful part here—and one that is central to Payload—is... that it's just JavaScript. No magic, no "do plugins the Payload way"... So you definitely would never have to fight a complex plugin infrastructure.

**This!** I think that's one of the biggest advantage that payload offers; there's so little overhead.

>For example, NextJS configs use a nice pattern that we can mimic now with Payload, like this:

>```
>const config = withPlugin({
>serverURL: 'http://localhost:3000',
>    collections: [
>    ],
> // the rest of the config
> })
>```

So if I do understand you correctly, this already works? Would it be possible for you to add an example like that in the docs or here? I'm still trying to wrap my head around the different concepts.

Or is it just really object.assign() this to the payload config?


>In the meantime, this Discussion would make for a great spot to centralize plugin ideas, as the community has them, so we can evaluate what's necessary to make them happen. If you have any ideas, send them our way and let's make it happen!

I think this also sounds great. So far I can't really say I have anything in mind, but I plan to at least try to build a small demo shop with it, so I think something as a starting point for ecommerce would be a great incentive for people to try payload imo.

Last thought: As payload is still pretty new I think it would be a good idea to start now with something like "awesome-payload" so people can use community packages / configs or at least have a currated way to find them.

**Reply by denolfe** [reply]
Correct, since Payload's config is just JS, as long as it passes the validation when being passed into the `buildConfig` function - it will work. The `PayloadConfig` type is what is expected by that function, which may be good to reference: https://github.com/payloadcms/payload/blob/master/src/config/types.ts#L63
